docs/design/sum_types.md

```diff
@@ -0,0 +1,200 @@
+# Sum types
+
+<!--
+Part of the Carbon Language project, under the Apache License v2.0 with LLVM
+Exceptions. See /LICENSE for license information.
+SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+-->
+
+<!-- toc -->
+
+## Table of contents
+
+-   [Overview](#overview)
+-   [`choice` declarations](#choice-declarations)
+-   [User-defined sum types](#user-defined-sum-types)
+-   [Alternatives considered](#alternatives-considered)
+-   [References](#references)
+
+<!-- tocstop -->
+
+## Overview
+
+In Carbon, a _sum type_ is a type whose values are grouped into several distinct
+named cases, called _alternatives_. A value of a sum type logically consists of
+a _discriminator_ tag that identifies which alternative is present, together
+with that alternative's value if it has one. Sum types are typically handled
+with pattern matching.
+
+## `choice` declarations
+
+The `choice` keyword is used to declare a sum type by specifying its interface,
+leaving the implementation to the compiler. A `choice` declaration consists of
+the `choice` keyword followed by the name of the type, and then a list of
+alternatives inside curly braces. An alternative declaration consists of the
+alternative name, optionally followed by a parameter list in parentheses. If
+present, the parameter list has the same syntax as in a
+[function declaration](README.md#functions). For example:
+
+```carbon
+choice OptionalI32 {
+  Some(value: i32),
```

Comment:
Are the parameter names used solely for documentation purposes? If so, I'd find it useful to say that somewhere near here. (And if not, I'd be interested in what else they do.)